Add quick pick support to infer sub expressions in extract to constant code action

TharushiJay opened the pull request:
## Purpose
This PR adds support for quick picks in the VSCode plugin to identify the sub expression selected by a user when extracting a complex expression to a constant using the `extract to constant` code action.

PR to the VSCode plugin : https://github.com/wso2-enterprise/ballerina-plugin-vscode/pull/559

Fixes #37480

## Samples

https://user-images.githubusercontent.com/27485094/195769926-b69cacb8-5d9b-467d-b75c-7a65e9f64545.mov

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

LakshanWeerasinghe commented on language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractToConstantCodeAction.java:
```diff
@@ -92,28 +100,72 @@ public List<CodeAction> getCodeActions(CodeActionContext context,
+        
+        LinkedHashMap<String, List<TextEdit>> textEditMap = new LinkedHashMap<>();
+        nodeList.forEach(extractableNode -> {
+            String value = extractableNode.toSourceCode().strip();
+            LineRange replaceRange = extractableNode.lineRange();
+            String constDeclStr = String.format("const %s %s = %s;%n", typeSymbol.get().signature(), constName, value);
+            TextEdit constDeclEdit =
+                    new TextEdit(new Range(constDeclPosition, constDeclPosition), constDeclStr);
+            TextEdit replaceEdit = new TextEdit(new Range(PositionUtil.toPosition(replaceRange.startLine()),
+                    PositionUtil.toPosition(replaceRange.endLine())), constName);
```
```suggestion

private List<TextEdit> getTextEdits(NonTerminalNode node, TypeSymbol typeSymbol, String constDeclStr {
             String value = node.toSourceCode().strip();
            LineRange replaceRange = node.lineRange();
            String constDeclStr = String.format("const %s %s = %s;%n", typeSymbol.get().signature(), constName, value);
            TextEdit constDeclEdit = new TextEdit(new Range(constDeclPosition, constDeclPosition), constDeclStr);
            TextEdit replaceEdit = new TextEdit(new Range(PositionUtil.toPosition(replaceRange.startLine()),
                    PositionUtil.toPosition(replaceRange.endLine())), constName);
                    return List.of(constDeclEdit, replaceEdit);
                    }
```

We can extract this code snippet to something like this. Where we can reuse it between 114-119 lines.

LakshanWeerasinghe commented on language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/ExtractToConstantWithQuickPickTest.java:
```diff
@@ -0,0 +1,73 @@
+/**
+ * Test cases for Extract to Constant code action with quick pick.
+ * 
+ * @since 2201.3.0
```
This should be `2201.3.1` or `2201.4.0`.

LakshanWeerasinghe reviewed (changes requested):
Also add test case for to check for `ObjectFieldNodes`.

TharushiJay commented:
> Also add test case for to check for ObjectFieldNodes.

> @TharushiJay shall we rebase this PR and address review comments?

Addressed the review comments and rebased the PR.

malinthar commented on language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractToConstantCodeAction.java:
```diff
@@ -91,30 +98,61 @@ public List<CodeAction> getCodeActions(CodeActionContext context,
+        // Selection is a position
+        List<Node> nodeList = new ArrayList<>();
+        // Identify the sub-expressions to be extracted
+        while (!(node instanceof StatementNode) && !(node instanceof ModuleMemberDeclarationNode)
```
Shall we move this validation to a separate private method.  Something like `isExpressionNode` ?

malinthar commented on language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractToConstantCodeAction.java:
```diff
@@ -91,30 +98,61 @@ public List<CodeAction> getCodeActions(CodeActionContext context,
+                && node.kind() != SyntaxKind.OBJECT_FIELD && !nodeValidator.getInvalidNode()) {
+            nodeList.add(node);
+            node = node.parent();
+            node.accept(nodeValidator);
```
Let's add a null check here. There is a possibility of parent() becoming null

malinthar commented on language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractToConstantCodeAction.java:
```diff
@@ -91,30 +98,61 @@ public List<CodeAction> getCodeActions(CodeActionContext context,
+        }
+        
+        // Selection is a position
+        List<Node> nodeList = new ArrayList<>();
+        // Identify the sub-expressions to be extracted
+        while (!(node instanceof StatementNode) && !(node instanceof ModuleMemberDeclarationNode)
+                && node.kind() != SyntaxKind.OBJECT_FIELD && !nodeValidator.getInvalidNode()) {
+            nodeList.add(node);
+            node = node.parent();
+            node.accept(nodeValidator);
+        }
```
Let's move this logic to a separate method called `getPossibleExpressionNodes`

malinthar commented on language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/AbstractCodeActionTest.java:
```diff
@@ -422,6 +422,32 @@ protected boolean validateAndModifyArguments(JsonObject actualCommand,
                 }
             }
             return false;
+        } else if ("ballerina.action.extract".equals(actualCommand.get("command").getAsString())) {
+            if (actualArgs.size() == 3) {
+                String actualName = actualArgs.get(0).getAsString();
+                String expectedName = expArgs.get(0).getAsString();
+
+                String actualFilePath = actualArgs.get(1).getAsString().replace(sourceRoot.toString(), "");
+                if (actualFilePath.startsWith("/") || actualFilePath.startsWith("\\")) {
+                    actualFilePath = actualFilePath.substring(1);
+                }
+                String expectedFilePath = expArgs.get(1).getAsString();
+                
+                JsonObject actualTextEdits = actualArgs.get(2).getAsJsonObject();
+                JsonObject expectedTextEdits = expArgs.get(2).getAsJsonObject();
+
+                if (actualName.equals(expectedName) && actualFilePath.equals(expectedFilePath)
+                        && actualTextEdits.equals(expectedTextEdits)) {
+                    return true;
+                }
+
+                JsonArray newArgs = new JsonArray();
+                newArgs.add(actualName);
+                newArgs.add(actualFilePath);
+                newArgs.add(actualTextEdits);
+                actualCommand.add("arguments", newArgs);
+            }
+            return false;
```
This method can grow very fast if we are to add each command. If this is common for each `extract to constant`, `extract to local var` and `extract to function` we can keep this. If not, shall we override this method in the respective child class?

TharushiJay replied:
This is common to all three code actions, hence keeping this.

malinthar commented on language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/AbstractCodeActionTest.java:
```diff
@@ -422,6 +422,32 @@ protected boolean validateAndModifyArguments(JsonObject actualCommand,
+                String expectedFilePath = expArgs.get(1).getAsString();
+                
+                JsonObject actualTextEdits = actualArgs.get(2).getAsJsonObject();
+                JsonObject expectedTextEdits = expArgs.get(2).getAsJsonObject();
```
Shall we validate the list of quick picks as well?

malinthar replied:
Resolving this comment as this is already done

malinthar reviewed (changes requested):
The changes look good! Let's see if the change requests can be addressed

mohanvive commented:
@TharushiJay can you please check the build failure?

TharushiJay commented:
> @TharushiJay can you please check the build failure?

Fixed.

IMS94 reviewed (approved):
LGTM!
